### PR TITLE
Feature/rlp 917/payment poc with comms

### PR DESCRIPTION
**Pull request by marcosmartinez7**
# Description
This PR includes the modifications to make a p2p payment between two lumino full nodes.

# Change log: 

- Removed is_subscribed_to (WARNING: this must be added again when it works)
- Moved validate_and_parse_message to transport/utils and renamed to validate_and_parse_message

# Issues: 
- some times lumino nodes dosnt receive messages or something, the rif comms node shows the msgs.

## Case 1: 
- Fresh start (start both comms node, start lumino nodes): OK

## Case 2:
- Shut down lumino nodes after case 1 succeeds and start them again. Try to send a payment: OK

## Case 3:
- Shut down rif comms and start them again (recovery from rif comms failures), try to send a payment: FAIL, there is no log on rif comms node, there are not subscribe invocations and it should. Why? thats because the send_message isnt invoked. The retry queue checks for the stop event of the transport and somehow it is stopped. 

Basically the streams are closed therefore the exception is raised and the transport is stoped

```
<_MultiThreadedRendezvous of RPC that terminated with:
	status = StatusCode.UNKNOWN
	details = "Stream removed"
	debug_error_string = "{"created":"@1607006465.758633539","description":"Error received from peer ipv4:127.0.0.1:5013","file":"src/core/lib/surface/call.cc","file_line":1061,"grpc_message":"Stream removed","grpc_status":2}"
>
```

We can work on that on https://jirainfuy.atlassian.net/browse/RLP-927

# TODO: 
- modify validate and parse message in order to just parse to string and make the validation outside (_handle_message)

## Builds on open PR #162

### CODE REVIEW

File: transport/matrix/utils.py [hunk 1]
```diff
@@ -21,12 +20,15 @@
...
+from binascii import Error as DecodeError
 from cachetools import LRUCache, cached
 from ecies import decrypt
 from eth_utils import decode_hex, encode_hex, to_canonical_address, to_normalized_address
 from gevent.event import Event
 from gevent.lock import Semaphore
 from matrix_client.errors import MatrixError, MatrixRequestError
+from raiden_contracts.constants import ID_TO_NETWORKNAME
+
 from raiden.exceptions import InvalidProtocolMessage, InvalidSignature, TransportError
```
Comment (mortelli) [inline transport/matrix/utils.py]: multiple unused imports here (up til line `40`)

File: transport/utils.py [hunk 2]
```diff
@@ -13,6 +16,14 @@
 from transport.node import Node as TransportNode
 from transport.udp import utils as udp_utils
 
+import structlog
+log = structlog.get_logger(__name__)
```
Comment (mortelli) [inline transport/utils.py:24]: please move this assignment below the imports section, and then apply reformat+optimize

File: transport/utils.py [hunk 3]
```diff
@@ -186,3 +197,73 @@ def __str__(self):
 
     def __repr__(self):
         return f"<{self.__class__.__name__} for {to_normalized_address(self.recipient)}>"
+
+
+def validate_and_parse_messages(data, peer_address) -> List[Message]:
```
Comment (mortelli) [inline transport/utils.py:201]: let's add a docstring for this function

File: transport/utils.py [hunk 4]
```diff
@@ -186,3 +197,73 @@ def __str__(self):
...
+                )
+                continue
+            if not isinstance(message, SignedMessage):
+                log.warning(
+                    "ToDevice Message not a SignedMessage!",
+                    message=message,
+                    peer_address=pex(peer_address),
+                )
+                continue
+            # TODO message must be signed by sender
```
Comment (mortelli) [inline transport/utils.py:273]: when we deal with this, **please** let's iterate on this method.

its structure leaves much to be desired, which is going to impact on its maintainability and readability

File: transport/rif_comms/node.py [hunk 5]
```diff
@@ -83,16 +83,17 @@ def _receive_messages(self):
...
         for notification in self._our_topic_stream:
-            raiden_message = self._notification_to_message(notification.channelNewData)
-            if raiden_message:
-                self.log.info("incoming message", message=raiden_message)
-                self._handle_message(raiden_message)
+            raiden_messages = self._notification_to_messages(notification.channelNewData)
+            if raiden_messages:
+                self.log.info("incoming messages", messages=raiden_messages)
+                for raiden_message in raiden_messages:
+                    self._handle_message(raiden_message)
```
Comment (mortelli) [inline transport/rif_comms/node.py]:
```suggestion
            for raiden_message in self._notification_to_messages(notification.channelNewData):
                self.log.info("incoming message", message=raiden_messages)
                self._handle_message(raiden_message)
```

Comment (mortelli) [reply transport/rif_comms/node.py]: returning `[]` (instead of `None`) on `_notification_to_messages` will do away with the problem of attempting to iterate on `NoneType`

File: transport/rif_comms/node.py [hunk 6]
```diff
@@ -108,10 +109,13 @@ def _notification_to_message(notification_data: ChannelNewData) -> RaidenMessage
...
             # the message is inside the notification data, encoded by the RIF Comms GRPC api
-            message_string = bytes(content["data"]).decode()
-            print("Message received", message_string)
-            message_dict = json.loads(message_string)
-            return message_from_dict(message_dict)
+            comms_message_string = bytes(content["data"]).decode()
+            # a comms message can contain one or N raiden messages, therefore it must be parsed
+            messages = validate_and_parse_messages(comms_message_string, None)
+            if not messages:
+                return None
```
Comment (mortelli) [inline transport/rif_comms/node.py]: why are we returning `None` when we expect a list?

my proposal:
```python
    # line 111
    comms_message_string = bytes(content["data"]).decode()
    # a comms message can contain one or N raiden messages, therefore it must be parsed
    return validate_and_parse_messages(comms_message_string, None)
return []
```

**Review by mortelli — COMMENTED** [review]
be sure to apply _Reformat Code_ and _Optimize Imports_ to all affected files

File: transport/matrix/utils.py (hunk 1 above)
Comment (marcosmartinez7) [reply transport/matrix/utils.py]: fixed

File: transport/utils.py (hunk 2 above)
Comment (marcosmartinez7) [reply transport/utils.py:24]: fixed

File: transport/utils.py (hunk 3 above)
Comment (marcosmartinez7) [reply transport/utils.py:201]: done

File: transport/utils.py (hunk 4 above)
Comment (marcosmartinez7) [reply transport/utils.py:273]: there is already an issue for this, sure

File: transport/rif_comms/node.py (hunk 6 above)
Comment (marcosmartinez7) [reply transport/rif_comms/node.py]: agree

File: raiden/tests/unit/rif_comms/test_rif_comms_client.py [hunk 7]
```diff
@@ -12,7 +12,7 @@
     },
```
Comment (mortelli) [inline raiden/tests/unit/rif_comms/test_rif_comms_client.py]: please undo **all** the changes to this file

File: transport/matrix/utils.py [hunk 8]
```diff
@@ -1,6 +1,5 @@
 import json
```
Comment (mortelli) [inline transport/matrix/utils.py:1]: this file needs to have `Optimize Imports` reapplied to it

File: transport/rif_comms/node.py [hunk 9]
```diff
@@ -1,10 +1,10 @@
-import json
 from typing import Any, Dict
 
 import structlog
 from eth_utils import is_binary_address
 from gevent import killall, wait
 from greenlet import GreenletExit
+
```
Comment (mortelli) [inline transport/rif_comms/node.py:7]: `Optimize Imports` will remove this nl

**Review by mortelli — APPROVED** [review]
please take care of these before merging

File: raiden/tests/unit/rif_comms/test_rif_comms_client.py (hunk 7 above)
Comment (marcosmartinez7) [reply raiden/tests/unit/rif_comms/test_rif_comms_client.py]: done

File: transport/matrix/utils.py (hunk 8 above)
Comment (marcosmartinez7) [reply transport/matrix/utils.py:1]: i suppose we have different pycharm config or smthng

File: transport/rif_comms/node.py (hunk 9 above)
Comment (marcosmartinez7) [reply transport/rif_comms/node.py:7]: same
